File: platform/platform-impl/src/com/intellij/diagnostic/SamplingTask.java
// Copyright 2000-2021 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.intellij.diagnostic;

import com.sun.management.OperatingSystemMXBean;
import kotlinx.coroutines.Job;
import org.jetbrains.annotations.NotNull;

import java.lang.management.GarbageCollectorMXBean;
import java.lang.management.ManagementFactory;
import java.lang.management.ThreadInfo;
import java.lang.management.ThreadMXBean;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;

class SamplingTask {
  private final static ThreadMXBean THREAD_MX_BEAN = ManagementFactory.getThreadMXBean();
  private final static List<GarbageCollectorMXBean> GC_MX_BEANS = ManagementFactory.getGarbageCollectorMXBeans();

  private final int myDumpInterval;
  private final int myMaxDumps;

  private final List<ThreadInfo[]> myThreadInfos = new ArrayList<>();

  private final Job job;

  private final long myStartTime;
  private long myCurrentTime;
  private final long myGcStartTime;
  private long myGcCurrentTime;
  private final double myProcessCpuLoad;

  SamplingTask(int intervalMs, int maxDurationMs) {
    myDumpInterval = intervalMs;
    myMaxDumps = maxDurationMs / intervalMs;
    myCurrentTime = myStartTime = System.nanoTime();
    myGcCurrentTime = myGcStartTime = currentGcTime();
    myProcessCpuLoad = ((OperatingSystemMXBean)ManagementFactory.getOperatingSystemMXBean()).getProcessCpuLoad();
    job = PerformanceWatcher.getInstance().scheduleWithFixedDelay(this::dumpThreads, myDumpInterval);
  }

  private void dumpThreads() {
    myCurrentTime = System.nanoTime();
    myGcCurrentTime = currentGcTime();
    ThreadInfo[] infos = ThreadDumper.getThreadInfos(THREAD_MX_BEAN, false);
    if (!job.isCancelled()) {
      myThreadInfos.add(infos);
      if (myThreadInfos.size() >= myMaxDumps) {
        stop();
      }
      dumpedThreads(ThreadDumper.getThreadDumpInfo(infos, true));
    }
  }

  protected void dumpedThreads(@NotNull ThreadDump threadDump) {
  }

  private static long currentGcTime() {
    return GC_MX_BEANS.stream().mapToLong(GarbageCollectorMXBean::getCollectionTime).sum();
  }

  public final int getDumpInterval() {
    return myDumpInterval;
  }

  public final List<ThreadInfo[]> getThreadInfos() {
    return myThreadInfos;
  }

  public final long getTotalTime() {
    return TimeUnit.NANOSECONDS.toMillis(myCurrentTime - myStartTime);
  }

  public final long getGcTime() {
    return myGcCurrentTime - myGcStartTime;
  }

  public final double getProcessCpuLoad() {
    return myProcessCpuLoad;
  }

  public final boolean isValid(long dumpingDuration) {
    return myThreadInfos.size() >= Math.max(10, Math.min(myMaxDumps, dumpingDuration / myDumpInterval / 2));
  }

  public void stop() {
    job.cancel(null);
  }
}

File: platform/platform-impl/src/com/intellij/diagnostic/SamplingTask.kt
// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.diagnostic

import com.intellij.util.containers.UList
import com.sun.management.OperatingSystemMXBean
import kotlinx.coroutines.*
import java.lang.management.ManagementFactory
import java.lang.management.ThreadInfo
import java.util.concurrent.TimeUnit
import kotlin.coroutines.coroutineContext
import kotlin.time.Duration.Companion.milliseconds

internal open class SamplingTask(@JvmField internal val dumpInterval: Int, maxDurationMs: Int, coroutineScope: CoroutineScope) {
  private val maxDumps: Int = maxDurationMs / dumpInterval

  var threadInfos: UList<Array<ThreadInfo>> = UList()
    private set

  private val job: Job?
  private val startTime: Long = System.nanoTime()
  private var currentTime: Long = startTime
  private val gcStartTime: Long = currentGcTime()
  private var gcCurrentTime: Long = gcStartTime
  val processCpuLoad: Double = (ManagementFactory.getOperatingSystemMXBean() as OperatingSystemMXBean).processCpuLoad

  val totalTime: Long
    get() = TimeUnit.NANOSECONDS.toMillis(currentTime - startTime)
  val gcTime: Long
    get() = gcCurrentTime - gcStartTime

  init {
    job = coroutineScope.launch {
      val delayDuration = dumpInterval.milliseconds
      while (true) {
        dumpThreads(asyncCoroutineScope = coroutineScope)
        delay(delayDuration)
      }
    }
  }

  private suspend fun dumpThreads(asyncCoroutineScope: CoroutineScope) {
    currentTime = System.nanoTime()
    gcCurrentTime = currentGcTime()
    val infos = ThreadDumper.getThreadInfos(THREAD_MX_BEAN, false)
    coroutineContext.ensureActive()

    threadInfos = threadInfos.add(infos)
    if (threadInfos.size > maxDumps) {
      stopDumpingThreads()
      return
    }

    asyncCoroutineScope.launch {
      val rawDump = ThreadDumper.getThreadDumpInfo(infos, true)
      val dump = EventCountDumper.addEventCountersTo(rawDump)
      dumpedThreads(dump)
    }
      // don't schedule yet another dumpedThreads - wait for completion
      .join()
  }

  protected open suspend fun dumpedThreads(threadDump: ThreadDump) {}

  open fun stop() {
    job?.cancel()
  }

  open suspend fun stopDumpingThreads() {
    job?.cancelAndJoin()
  }
}

private val THREAD_MX_BEAN = ManagementFactory.getThreadMXBean()
private val GC_MX_BEANS = ManagementFactory.getGarbageCollectorMXBeans()

private fun currentGcTime(): Long = GC_MX_BEANS.sumOf { it.collectionTime }